### PR TITLE
cairo, bump version, include glib2

### PR DESCRIPTION
**Pull request by Begasus**
Should be ok, included glib2 to check the ability to test gobject-introspection ... (test on that still fails because it can't find cairo-gobject.h header) ...

### CODE REVIEW

File: x11-libs/cairo/patches/cairo-1.14.12.patchset [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ #if HAVE_FCFINI
+ #include <fontconfig/fontconfig.h>
+-- 
+2.10.2
+
+
+From 66cf163733ee56380b0525b918b7969849ed0229 Mon Sep 17 00:00:00 2001
+From: Jerome Duval <jerome.duval@gmail.com>
+Date: Fri, 21 Nov 2014 22:11:40 +0000
+Subject: link tests against libnetwork.
```
[the remark this answers is not in the document]
Comment (korli) [reply x11-libs/cairo/patches/cairo-1.14.12.patchset]: And how is it exactly better? How can it be upstreamed if it resides in the recipe?

File: x11-libs/cairo/cairo-1.14.12.recipe [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	devel:libfreetype$secondaryArchSuffix
+	devel:libglib_2.0$secondaryArchSuffix
+	devel:libpixman_1$secondaryArchSuffix
+	devel:libpng16$secondaryArchSuffix
+	devel:libz$secondaryArchSuffix
+	"
+BUILD_PREREQUIRES="
+	cmd:aclocal
+	cmd:autoconf
+	cmd:automake
```
[the remark this answers is not in the document]
Comment (Begasus) [reply x11-libs/cairo/cairo-1.14.12.recipe]: cmd:autorecon is part of the package called by cmd:autoconf ... (automake shouldn't be in here also) ...

File: x11-libs/cairo/cairo-1.14.12.recipe [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	MPL v1.1"
+REVISION="1"
+SOURCE_URI="http://cairographics.org/releases/cairo-$portVersion.tar.xz"
+CHECKSUM_SHA256="8c90f00c500b2299c0a323dd9beead2a00353752b2092ead558139bd67f7bf16"
+PATCHES="cairo-$portVersion.patchset"
+
+ARCHITECTURES="!x86_gcc2 x86 x86_64"
+SECONDARY_ARCHITECTURES="?x86"
+
+libVersion="2.11400.12"
```
[the remark this answers is not in the document]
Comment (Begasus) [reply x11-libs/cairo/cairo-1.14.12.recipe]: I don't (speek m4) :/

File: x11-libs/cairo/cairo-1.14.12.recipe [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	2002-2010 many others"
+LICENSE="GNU LGPL v2.1
+	MPL v1.1"
+REVISION="1"
+SOURCE_URI="http://cairographics.org/releases/cairo-$portVersion.tar.xz"
+CHECKSUM_SHA256="8c90f00c500b2299c0a323dd9beead2a00353752b2092ead558139bd67f7bf16"
+PATCHES="cairo-$portVersion.patchset"
+
+ARCHITECTURES="!x86_gcc2 x86 x86_64"
+SECONDARY_ARCHITECTURES="?x86"
```
[the remark this answers is not in the document]
Comment (Begasus) [reply x11-libs/cairo/cairo-1.14.12.recipe]: Haven't been able to build for x86_gcc2 yet ...

File: x11-libs/cairo/cairo-1.14.12.recipe [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+{
+	make install
+
+	rm $libDir/libcairo*.la
+
+	prepareInstalledDevelLibs libcairo libcairo-script-interpreter libcairo-gobject
+	fixPkgconfig
+
+	packageEntries devel \
+		$developDir
```
[the remark this answers is not in the document]
Comment (fbrosson) [reply x11-libs/cairo/cairo-1.14.12.recipe]: Well, this is a bit special, because we're not in `$docDir` but in `$dataDir`, so we can't move it to `$developDocDir`. (And there is no `developDataDir`.)

File: x11-libs/cairo/cairo-1.14.12.recipe (hunk 2 above)
[the remark this answers is not in the document]
Comment (fbrosson) [reply x11-libs/cairo/cairo-1.14.12.recipe]: Yep, but if the recipe calls `autoreconf` then it is recommended to mention it in the `BUILD_PREREQUIRES`.
This is safer in case autoreconf is moved to a subpackage or if, at some point in the future, an alternate package provides a `cmd:autoreconf`.

File: x11-libs/cairo/patches/cairo-1.14.12.patchset (hunk 1 above)
[the remark this answers is not in the document]
Comment (alaviss) [reply x11-libs/cairo/patches/cairo-1.14.12.patchset]: This patch just add `-lnetwork` unconditionally, and of course won't be accepted by upstream at all. If one wanted to upstream it, I would recommend fixing the build system so that `-lnetwork` is added automatically on Haiku.

IMO, unless the patch is improved so that it could be upstreamed, we should just drop it to reduce maintainance cost.

File: x11-libs/cairo/cairo-1.14.12.recipe (hunk 3 above)
[the remark this answers is not in the document]
Comment (alaviss) [reply x11-libs/cairo/cairo-1.14.12.recipe]: This formula is kinda complicated. It's perfectly valid (based on my guess) for 1.15.10 (unstable version), but is invalid for 1.14.12 (stable). So I think we should just ignore this (for now).

**Comment by Begasus** [comment]
Still fails to build for me for x86_gcc2 arch:

  CC       cairo-clip-polygon.lo
/sources/cairo-1.14.12/src/cairo-boxes-intersect.c: In function `intersect':
/sources/cairo-1.14.12/src/cairo-boxes-intersect.c:399: `out' redeclared as different kind of symbol
/sources/cairo-1.14.12/src/cairo-boxes-intersect.c:503: previous declaration of `out'
/sources/cairo-1.14.12/src/cairo-boxes-intersect.c:399: `out' redeclared as different kind of symbol
/sources/cairo-1.14.12/src/cairo-boxes-intersect.c:512: previous declaration of `out'
/sources/cairo-1.14.12/src/cairo-boxes-intersect.c:399: `out' redeclared as different kind of symbol
/sources/cairo-1.14.12/src/cairo-boxes-intersect.c:521: previous declaration of `out'
Makefile:2349: recipe for target 'cairo-boxes-intersect.lo' failed

File: x11-libs/cairo/cairo-1.14.12.recipe (hunk 2 above)
[the remark this answers is not in the document]
Comment (Begasus) [reply x11-libs/cairo/cairo-1.14.12.recipe]: This doesn't add up to the comments on https://github.com/haikuports/haikuports/pull/2230 ?

[the remark this answers is not in the document]
Comment (Begasus) [reply x11-libs/cairo/cairo-1.14.12.recipe]: So we should drop "autoconf, automake" here? Not sure you "can" remove aclocal ...

[the remark this answers is not in the document]
Comment (alaviss) [reply x11-libs/cairo/cairo-1.14.12.recipe]: Generally, you could just drop anything that's not called by the recipe. `autoconf`, `automake` are dependencies of `autoreconf`, so they would be pulled automatically.

[the remark this answers is not in the document]
Comment (fbrosson) [reply x11-libs/cairo/cairo-1.14.12.recipe]: I think it is better to keep in `BUILD_PREREQUIRES` most of these `cmd:*`  which are mentioned in configure.
Although `cmd:autoreconf` is (currently) provided by the same package which also provides `cmd:aclocal`, `cmd:autoconf` and `cmd:automake`, this could change in the future, so we should not, imho, drop these.

`cmd:makeinfo` can be dropped because it is not usefull for us. The only thing it does, if I'm not mistaken, is that it creates a `dir` file in `$infoDir`, and we don't like to have this `dir` file.

[the remark this answers is not in the document]
Comment (Begasus) [reply x11-libs/cairo/cairo-1.14.12.recipe]: I did some changes back, but got told to add/remove them, getting you guys in line would be nice for us to know if we should use them or not :)

[the remark this answers is not in the document]
Comment (fbrosson) [reply x11-libs/cairo/cairo-1.14.12.recipe]: To decide whether or not it would be OK to remove a `cmd:` from `BUILD_PREREQUIRES`, we need to know if the recipe would still build if we did remove that `cmd:` and also had **_each_** of the remaining `cmd:` items in a **_separate_** package or subpackage.
We should not assume that any set of commands will always be provided by the same package.
There is only one exception, and it's for the trivial commands like `cp`, `mv`, `rm`, `ln`, `mkdir`, `install` (and probably other very common commands), which are in coreutils.
BTW, commands like `tr`, `sed`, and maybe `cut` too, should be mentioned in `BUILD_PREREQUIRES` if used in the recipe or by the build processes.
In other words, it would only be OK to not mention a `cmd:` if it is absolutely obvious that it will allways be in coreutils.

[the remark this answers is not in the document]
Comment (fbrosson) [reply x11-libs/cairo/cairo-1.14.12.recipe]: > Generally, you could just drop anything that's not called by the recipe. `autoconf`, `automake` are dependencies of `autoreconf`, so they would be pulled automatically.

@alaviss: OK, I see what you mean. I have to admit that you are right.
But I still think that when we know that the configure script tests for the presence of aclocal, autoconf, autoheader, automake, then it is safer to mention the corresponding `cmd:` in `BUILD_PREREQUIRES`.

[the remark this answers is not in the document]
Comment (fbrosson) [reply x11-libs/cairo/cairo-1.14.12.recipe]: If I'm not mistaken, I already saw recipes which only call runConfigure and make, and which:
* fail to build if we don't mention at least some of these `cmd:`.
* build fine if we do mention at least a minimal set of `cmd:`.

Consider `cmd:pkg_config`. It is never called directly from the recipe. If we chose to only mention those `cmd:` that we call from the recipe, then we would "have" to drop `cmd:pkg_config`. We don't do that because we know that the build process needs it.

So here, if we know that some tools are needed (or even merely tested for their presence) during the build, then we should mention them in `BUILD_PREREQUIRES`.

@alaviss (and others), what do you think?
Would you agree to keep the `cmd:aclocal`, `cmd:autoconf`, `cmd:autoheader` and `cmd:automake` even if the recipe only calls autoreconf?

[the remark this answers is not in the document]
Comment (fbrosson) [reply x11-libs/cairo/cairo-1.14.12.recipe]: When haikuporter sees the `cmd:autoreconf` in `BUILD_PREREQUIRES`, it searches for a package providing it, and "activates" it in the chroot. Now, since it is currently provided by the autoconf package, the chroot will automatically also have all the other `cmd:` provided by that package: `cmd:autoconf`, `cmd:autoheader`, `cmd:autom4te`, `cmd:autoscan`, `cmd:autoupdate` and `cmd:ifnames`.
So, even if you omit these commands, they will be available in the chroot if you mention `cmd:autoreconf`. And you might not even notice that some of them are called.
The problem is that if, at some point in the future, the layout of autoconf package is changed and some commands moved to subpackages or removed from that package, then they would no longer be "pulled" by `cmd:autoreconf`.
That's why, imho, it is safer to _also_ mention `cmd:autoconf` and `cmd:autoheader`, instead of just `cmd:aclocal`, `cmd:automake` and `cmd:autoreconf`.

**Comment by pulkomandy** [comment]
> When haikuporter sees the `cmd:autoreconf` in `BUILD_PREREQUIRES`, it searches for a package providing it, and "activates" it in the chroot. Now, since it is currently provided by the autoconf package, the chroot will automatically also have all the other `cmd:` provided by that package: `cmd:autoconf`, `cmd:autoheader`, `cmd:autom4te`, `cmd:autoscan`, `cmd:autoupdate` and `cmd:ifnames`.
> So, even if you omit these commands, they will be available in the chroot if you mention `cmd:autoreconf`. And you might not even notice that some of them are called.
> The problem is that if, at some point in the future, the layout of autoconf package is changed and some commands moved to subpackages or removed from that package, then they would no longer be "pulled" by `cmd:autoreconf`.
> That's why, imho, it is safer to _also_ mention `cmd:autoconf` and `cmd:autoheader`, instead of just `cmd:aclocal`, `cmd:automake` and `cmd:autoreconf`.

I think we don't need a strict rule here and rather use common sense
(and accept that it will vary depending on who writes the recipe).

As I see it:
- If the recipe explicitly calls something, it should be declared
- If the buildsystem calls something, ideally it should also be declared
  (for example if a configure script or makefile needs sed). But it
  would be very annoying to survey the whole buildsystem to know
  everything that is run.
- However, if something is not called directly by either the recipe or
  the buildsystem (for example, automake needs m4, even if it is not
  directly invoked by the buildscript), then this should rather be
  handled by having the automake package require cmd:m4.

So, even if we could move aclocal, autoheader, ... outside the autoconf
package, we would likely make autoconf depend on the new packages
anyways. So there is no need to list everything in each recipe. However,
if calling aclocal directly in the recipe, it is wise to add it to the
requires.

**Comment by fbrosson** [comment]
> But it would be very annoying to survey the whole buildsystem to know everything that is run.

I agree with you.
BTW, the only thing I do when I want to check if I should mention `cmd:aclocal`, `cmd:autoconf`, `cmd:autoheader` and `cmd:automake` is that I look for these strings in configure: ACLOCAL, AUTOCONF, AUTOHEADER and AUTOMAKE. That's all I do. I don't even try to see what other tools could be needed. If a recipe builds fine, then I know that I have probably all the required `cmd:` in `BUILD_PREREQUIRES`.

And, as you say, some contributors can choose to only mention `cmd:autoreconf`, while others might like to also mention `cmd:autoconf` and `cmd:autoheader`. I recommend this second option, but I don't object the first option.

**Comment by Begasus** [comment]
I'm not arguing if it's needed or not, it's confusing that (in the past) I've been told to sometime use it and sometimes not (even though I've seen the cmd's passing by) ...

**Comment by fbrosson** [comment]
No problem :)
I like to mention `cmd:autoconf` & `cmd:autoheader` for the reasons explained above, but I'll only do so in  new recipes. And I won't ask anyone to add these because it's not strictly required.
So please feel free to not add these if you wish. I'm OK with this and I won't complain :)

**Comment by fbrosson** [comment]
@alaviss, I've been thinking about what you wrote several days ago (regarding the fact that we should drop the unneeded `cmd:autoconf` and `cmd:autoheader` in `BUILD_PREREQUIRES`) and I wanted to tell you that **you have convinced me**. So thanks, @alaviss & @pulkomandy for your comments, and sorry, @Begasus, for my wrong suggestions.

From now on, in recipes that call autoreconf, I won't add `cmd:autoconf` and `cmd:autoheader` anymore. I'll only make sure we have `cmd:aclocal`, `cmd:automake` and `cmd:autoreconf`.
I agree that if we ever move `cmd:autoreconf` from the autoconf package to a subpackage (or to another package) then that target (sub)package will have to declare a dependency on `cmd:autoconf` and `cmd:autoheader`.
(Apologies to you all, btw :)

**Comment by fbrosson** [comment]
:bomb: I tried the tests (on x86_64) and it seemed to work... until the system became unstable and KDLed.
Maybe this is only on my system, but I'd recommend to testing this recipe a bit more before any merge.

BTW, @alaviss, I think it would be better to only approve a PR if you have tested it. IMHO the "approve" action should only be used when you have really tested the recipe. Otherwise a simple "comment" review is enough :)

**Comment by Begasus** [comment]
Mentioned it on IRC, needed to add it here too :) sorry @fbrosson 
Before the system froze on me (which wouldn't be a first one) I was able to save a debug report on an error, after that the test went through untill the freeze ...
https://pastebin.com/BTeHwGEF

**Comment by mmlr** [comment]
And again: Please report all KDLs to the Haiku bugtracker.

**Comment by Begasus** [comment]
Can't report a KDL if I didn't encounter one (yet)

**Comment by fbrosson** [comment]
@Begasus: No problem. (BTW, I should be on IRC...)

**Comment by Begasus** [comment]
I was at work :)

**Comment by Begasus** [comment]
Ran into this (earlier and today too): https://pastebin.com/cXLg1F12

**Comment by waddlesplash** [comment]
Superseded by https://github.com/haikuports/haikuports/commit/fb1ac637d8c1900aa429625acbcc5137bd908c43.
